mypy/semanal.py

```diff
@@ -3830,8 +3830,8 @@ def visit_star_expr(self, expr: StarExpr) -> None:
             expr.expr.accept(self)
 
     def visit_yield_from_expr(self, e: YieldFromExpr) -> None:
-        if not self.is_func_scope():  # not sure
-            self.fail('"yield from" outside function', e, serious=True, blocker=True)
+        if not self.is_func_scope() or self.is_comprehension_stack[-1]:
+            self.fail('"yield from" outside function or lambda', e, serious=True, blocker=True)
```

Comment:
I don't think we need the "or lambda" part. Lambdas are functions, and it's very rare to have a yield in a lambda, so it seems more likely to confuse users than to help them.

Reply:
Also, it would be good to have a separate error message for yields in comprehensions.

Reply:
Ok